posts/inside-rust/2021-03-15-What-the-error-handling-project-group-is-working-towards.md

```diff
@@ -0,0 +1,201 @@
+
+If you're interested in more details we've created a gist of the proof of concept [here](https://gist.github.com/yaahc/c9348cca4aa34ba8bc92f41ba543919d).
+
+The main difference is that we hard code the `From` impl into the new `Result` type, rather than having it be inherited from the wrapped error types the way `std::result::Result` does.
+
+So instead of this:
+
+```rust
+impl<T, E, F> FromResidual<Result<!, E>> for Result<T, F>
+where
+    F: From<E> { ... }
+```
+
+We can write this:
+
+```rust
+impl<T, E> FromResidual<Result<!, E>> for MyResult<T>
+where
+    BoxError: From<E> { ... }
+```
+
+That way the error type associated with the custom result doesn't need to implement `From`, freeing it from the overlap rule and letting it implement `Error` instead.
+
+Assuming this works as well as we hope it does, and doesn't introduce worse problems than we're trying to fix, the plan is to then standardize this approach to all catch-all error types across the ecosystem. Our current vague plan has three parts. Add a new `Box<dyn Error>` equivalent to std that implement's `Error` and doesn't implement `From`, add a new `Result` type to `std` that doesn't have a generic parameter for `E`, and instead always converts to `YeetError`, or whatever it's called, whenever converting from a `std::result::Result`, and finally add lints to either `rustc` or `clippy` that warn against using `Box<dyn Error>`, and encourage people to switch to the new fixed equivalent type.
+
+
+### Disclaimers
+
+Note, we almost certainly won't _start_ with adding these types to std, and will instead trial this approach in 3rd party libraries, probably in a fork of `eyre` or `anyhow`. We're particularly worried about backwards compatibility so its easy to switch to the new libraries that properly implement `Error`. We're also worried about teachability, particularly with respect to knowing which `Result` type to use in each situation, and we're looking into what kind of errors get produced when you pick the wrong Error or Result type, and how we can guide users to the correct types.
+
+Also worth noting, this doesn't change how `std::result::Result` works. If you're happy with enum based error handling and `From` impls as necessary then you can still stick with the current `Result` type and `?` as you currently do. These changes only apply to types like `Box<dyn Error>` that implement From for all error types so they can still implement error while using `?` for convenience.
+
+## Summary
+
+To summarize on the above our plan is to:
+
+1. Move the error trait into core
```

Comment:
This comes a bit out of nowhere considering this is the first item on the plan. There was no other mention of core in the rest of the blog post.